service/store/server/blob.go

```
@@ -0,0 +1,148 @@
+	} else if err != nil {
+		return errors.InternalServerError("store.Blob.Read", err.Error())
+	}
+
+	// execute the request
+	blob, err := store.DefaultBlobStore.Read(req.Key, gostore.BlobNamespace(req.Options.Namespace))
+	if err == gostore.ErrNotFound {
+		return errors.NotFound("store.Blob.Read", "Blob not found")
+	} else if err == gostore.ErrMissingKey {
+		return errors.BadRequest("store.Blob.Read", "Missing key")
+	} else if err != nil {
+		return errors.InternalServerError("store.Blob.Read", err.Error())
+	}
+
+	// read from the blob and stream it to the client
+	buffer := make([]byte, bufferSize)
+	for {
+		if _, err := blob.Read(buffer); err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		if err := stream.Send(&pb.BlobReadResponse{Blob: buffer}); err != nil {
+			return err
+		}
+	}
+
+	return stream.Close()
+}
+
+func (b *blobHandler) Write(ctx context.Context, stream pb.BlobStore_WriteStream) error {
+	// the key and options are passed on each message but we only need to extract them once
+	var buf *bytes.Buffer
+	var key string
+	var options *pb.BlobOptions
+
+	// recieve the blob from the client
+	for {
+		req, err := stream.Recv()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return errors.InternalServerError("store.Blob.Write", err.Error())
+		}
+
+		if buf == nil {
+			// first message recieved from the stream
+			buf = bytes.NewBuffer(req.Blob)
+			key = req.Key
+			options = req.Options
+		} else {
+			// subsequent message recieved from the stream
+			buf.Write(req.Blob)
+		}
+	}
+
+	// ensure the blob was sent over the stream
+	if buf == nil {
+		return errors.BadRequest("store.Blob.Write", "No blob was sent")
+	}
+
+	// parse the options
+	if options == nil || len(options.Namespace) == 0 {
+		options = &pb.BlobOptions{Namespace: namespace.FromContext(ctx)}
+	}
+
+	// authorize the request
+	if err := authns.Authorize(ctx, options.Namespace); err == authns.ErrForbidden {
```

anyway we can do this authorization before receiving the data?

---

I couldn't find any way, because we pass the namespace etc in the options and that is part of the request payload. I could rely on passing it via the req headers but that feels hacky.

---

Thought as much. I wonder if we have to pull it up in to the loop where you're assigning to `options` . From an end user's POV it will give the best experience (e.g. fail fast). Admittedly looks a bit fugly from a code perspective